### PR TITLE
Terraform workflow skeleton

### PR DESCRIPTION
**Pull request by samrabelachew**
Created a skeleton of the terraform workflow structure. Subsequent PRs will handle the TODOs and build out its true functionality. As is:

- defines the workflow, spins up a temporal worker
- executes the terraform workflow per revision as a child workflow execution
- defines the activities that will operate in each execution
- wraps execution with a session

### CODE REVIEW

File: server/neptune/workflows/internal/deploy/terraform.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+// Export anything that callers need such as requests, signals, etc.
+type TerraformRequest = terraform.Request
+
+var TerraformTaskQueue = terraform.TaskQueue
+
+// To be able to execute a child workflow within the deploy workflow and avoid import cycles, terraform workflow
+// is defined within deploy package
+func Terraform(ctx workflow.Context, request TerraformRequest) error {
+	return terraform.Workflow(ctx, request)
+}
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/deploy/terraform.go:17]: Why are there import cycles? we shouldn't be referencing things in the deploy package from the terraform package? If we are let's move them, I would rather us keep a clean structure of one workflow per package.

File: server/neptune/workflows/internal/deploy/workflow.go [hunk 2]
```diff
@@ -55,7 +58,55 @@ func Workflow(ctx workflow.Context, request Request) error {
...
+			Plan: terraform.Job{
+				Steps: convertToTerraformSteps(request.Root.Plan.Steps),
+			},
+		},
+	}
+	err = workflow.ExecuteChildWorkflow(ctx, terraform.Workflow, terraformWorkflowRequest).Get(ctx, nil)
+	if err != nil {
+		return errors.Wrap(err, "executing child terraform workflow")
+	}
+	return nil
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/deploy/workflow.go:94]: Why is this here? this should be happening for each revision in the queue worker

File: server/neptune/workflows/internal/terraform/workflow.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+type PlanStatus int
+
+const (
+	Approved PlanStatus = iota
+	Rejected
+)
+
+func Workflow(ctx workflow.Context, request Request) error {
+	// TODO: should we set up a heartbeat? How long it takes to make progress on a terraform op activity is dependent on
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/terraform/workflow.go:21]: yes we should setup a heartbeat, something like every minute should be fine.

File: server/neptune/workflows/internal/terraform/workflow.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+		return err
+	}
+	return nil
+}
+
+type PlanReview struct {
+	Status PlanStatus
+}
+
+//TODO: should we build a timer select branches to notify users that an approval is still needed/way to cancel long running exection?
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/terraform/workflow.go:159]: we probably don't need it right now.

File: server/neptune/workflows/internal/terraform/workflow.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+func (r *Runner) WaitUntilPlanIsReviewed(ctx workflow.Context) PlanReview {
+	s := workflow.NewSelector(ctx)
+	var planReview PlanReview
+	signalChan := workflow.GetSignalChannel(ctx, "planreview-repo-root")
+	s.AddReceive(signalChan, func(c workflow.ReceiveChannel, more bool) {
+		c.Receive(ctx, &planReview)
+	})
+	s.Select(ctx)
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/terraform/workflow.go:167]: btw you don't need a select for just one branch, you can just call the channel's receive.

File: server/neptune/workflows/internal/activities/notify.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+type notifyActivities struct{}
+
+type NotifyRequest struct {
+	Operation TerraformOperation
+}
+
+type NotifyResponse struct {
+}
+
+func (t *terraformActivities) Notify(ctx context.Context, request NotifyRequest) error {
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/activities/notify.go:21]: wait what does this do?

File: server/neptune/workflows/internal/activities/execute.go [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package activities
+
+import "context"
+
+type executeCommandActivities struct{}
+
+type ExecuteCommandRequest struct {
+}
+
+func (t *terraformActivities) ExecuteCommand(ctx context.Context, request ExecuteCommandRequest) error {
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/activities/execute.go:10]: This is just for generic run commands yeah?

File: server/neptune/workflows/internal/activities/vcs.go [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package activities
+
+import (
+	"context"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/github"
+)
+
+// TODO: Initial implementation will support GH VCS, but we will need to eventually abstract to other platforms
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/activities/vcs.go:8]: we probably don't need to worry about abstracting for a long time so it's fine to focus on gh.

File: server/neptune/workflows/internal/terraform/workflow.go [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+func newRunner(ctx workflow.Context, request Request) *Runner {
+	return &Runner{
+		workerActivities: activities.Terraform{},
+		request:          request,
+	}
+}
+
+func (r *Runner) Run(ctx workflow.Context) error {
+	// Clone repository into disk
+	err := workflow.ExecuteActivity(ctx, r.GithubRepoClone, activities.GithubRepoCloneRequest{}).Get(ctx, nil)
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/terraform/workflow.go:70]: It would be great to have this such that we're only keeping the root directory on disk, instead of cloning the whole repo.  Even if we clone the whole repo, and extract the root directory and delete the rest, that would be better.

File: server/neptune/workflows/internal/terraform/workflow.go [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+			return errors.Wrap(err, "running step")
+		}
+	}
+	// Wait for plan review signal
+	planReview := r.WaitUntilPlanIsReviewed(ctx)
+	if planReview.Status == Rejected {
+		return nil
+	}
+	// Run apply steps
+	for _, step := range r.request.Root.Plan.Steps {
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/terraform/workflow.go:87]:
```suggestion
	for _, step := range r.request.Root.Apply.Steps {
```

File: server/neptune/workflows/internal/main.go [hunk 11]
```diff
@@ -0,0 +1 @@
+package internal
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/main.go:1]: ?

File: server/neptune/workflows/deploy.go [hunk 12]
```diff
@@ -3,16 +3,21 @@ package workflows
...
 type DeployRequest = deploy.Request
 type Repo = deploy.Repo
+type Root = deploy.Root
+type Job = deploy.Job
+type Step = deploy.Step
 
 type DeployNewRevisionSignalRequest = revision.NewRevisionRequest
 
 var DeployTaskQueue = deploy.TaskQueue
+var TerraformTaskQueue = terraform.TaskQueue
```
Comment (nishkrishnan) [inline server/neptune/workflows/deploy.go:20]: nit: I created `deploy.go` with the intention of having separate files for each workflow.

File: server/neptune/temporalworker/server.go [hunk 13]
```diff
@@ -104,39 +104,44 @@ func NewServer(config *Config) (*Server, error) {
...
 	defer s.Logger.Close()
 	defer s.TemporalClient.Close()
 
-	go func() {
-		defer wg.Done()
-		w := worker.New(s.TemporalClient, workflows.DeployTaskQueue, worker.Options{
+	// we create a base context that is marked done when we get a sigterm.
+	// we should use this context for other async work to ensure we
+	// are gracefully handling shutdown and not dropping data.
+	mainCtx, stop := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
```
Comment (nishkrishnan) [inline server/neptune/temporalworker/server.go:113]: sickkkkkk

File: server/neptune/temporalworker/server.go [hunk 14]
```diff
@@ -104,39 +104,44 @@ func NewServer(config *Config) (*Server, error) {
...
+			log.Fatalln("unable to start worker", err)
+		}
+		return nil
+	})
+	group.Go(func() error {
+		w := worker.New(s.TemporalClient, workflows.TerraformTaskQueue, worker.Options{
 			// ensures that sessions are preserved on the same worker
 			EnableSessionWorker: true,
 		})
 		if err := w.Run(worker.InterruptCh()); err != nil {
```
Comment (nishkrishnan) [inline server/neptune/temporalworker/server.go:130]: Hmm kinda weird that we're still using an interruptchannel here when we're keeping this in a shared error group with a context, but it seems like it would still work fine?

File: server/neptune/gateway/event/step_generator.go [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+)
+
+const defaultWorkspace = "default"
+
+type StepGenerator interface {
+	GeneratePlanSteps(repoID string) []workflows.Step
+	GenerateApplySteps(repoID string) []workflows.Step
+}
+
+type TerraformWorkflowStepsGenerator struct {
```
Comment (nishkrishnan) [inline server/neptune/gateway/event/step_generator.go:16]: Feels a bit premature to have this in a separate struct entirely from the push event handler, especially if all this is doing is just generating the steps.  

We're going to need the project config for figuring out the root definition/directory as well, not just here. I'd just keep this code in the push event handler for now.

**Review by nishkrishnan — COMMENTED** [review]
Looking good so far, I'm curious on what you think the success criteria of this task is at some point, I think we can call this good and have a number of smaller scoped tasks that you should feel free to breakdown for getting this workflow complete.

File: server/neptune/workflows/internal/activities/execute.go (hunk 7 above)
Comment (samrabelachew) [reply server/neptune/workflows/internal/activities/execute.go:10]: Yes, I figured because it's configurable to be any kind of generic run command, there isn't a guarantee if it is deterministic, so I made it an activity.

File: server/neptune/workflows/internal/main.go (hunk 11 above)
Comment (samrabelachew) [reply server/neptune/workflows/internal/main.go:1]: Go complained that there wasn't a go file within the internal package directly, so I just quickly threw in a main.go file to silence the error.

File: server/neptune/workflows/internal/deploy/terraform.go (hunk 1 above)
Comment (samrabelachew) [reply server/neptune/workflows/internal/deploy/terraform.go:17]: I think the import cycle was higher (deploy package shouldn't reference things defined in the internal package) like using an exported version of the terraform workflow request (`type TerraformRequest = terraform.Request`).

I can move it back to the original structure, but it just means the deploy package needs to be able to directly reference the terraform package. So really, we wouldn't need this file at all to act as the "external" representation of the workflow.

File: server/neptune/workflows/deploy.go (hunk 12 above)
Comment (samrabelachew) [reply server/neptune/workflows/deploy.go:20]: see discussion above on import cycles

File: server/neptune/workflows/internal/terraform/workflow.go (hunk 9 above)
Comment (samrabelachew) [reply server/neptune/workflows/internal/terraform/workflow.go:69]: Will look into it! on success criteria for PR, I basically want this skeleton passed without the implementation details of each activity and the initial push event trigger. I'll update those in separate PRs.

File: server/neptune/temporalworker/server.go (hunk 14 above)
Comment (samrabelachew) [reply server/neptune/temporalworker/server.go:130]: Yeah, the only other way to start a worker up was non-blocking: https://pkg.go.dev/go.temporal.io/sdk/worker#Worker. 

Wanted to test it a few times and see shutdown behavior in action. Alternatively will just revert back to regular go routines for this server.

Comment (samrabelachew) [reply server/neptune/temporalworker/server.go:126]: I reverted back to go routines. I don't think go groups fit our use case if we can already stop the workers by using the interrupt channel, and I didn't want the go group to exit in the case it handles the context cancel before the internal `w.Run()` call did.

File: server/neptune/temporalworker/server.go [hunk 16]
```diff
@@ -104,19 +104,27 @@ func NewServer(config *Config) (*Server, error) {
...
-		w := worker.New(s.TemporalClient, workflows.DeployTaskQueue, worker.Options{
+		w := worker.New(s.TemporalClient, workflows.TerraformTaskQueue, worker.Options{
 			// ensures that sessions are preserved on the same worker
 			EnableSessionWorker: true,
 		})
 		if err := w.Run(worker.InterruptCh()); err != nil {
-			log.Fatalln("unable to start worker", err)
+			log.Fatalln("unable to start terraform worker", err)
 		}
 	}()
```
Comment (nishkrishnan) [inline server/neptune/temporalworker/server.go:129]: hmm seems a bit pointless to run a separate worker process tbh.  it's not like we can scale this separately.  I'm wondering if we should just use the same deploy task queue/worker for now.

File: server/neptune/workflows/internal/deploy/revision/queue/worker.go [hunk 17]
```diff
@@ -102,14 +105,28 @@ func (w *Worker) work(ctx workflow.Context, revision string) error {
 	logger.Info(ctx, fmt.Sprintf("latest deployed revision %s", deployedRevision))
 
 	// TODO: fill in the rest
-
+	// Spin up a child workflow to handle Terraform operations
+	childWorkflowOptions := workflow.ChildWorkflowOptions{
+		TaskQueue: terraform.TaskQueue,
+		//TODO: match workflow id to format used by deploy workflow
+		WorkflowID: fmt.Sprintf("%s-%s", w.Repo.GetFullName(), w.Root.Name),
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/deploy/revision/queue/worker.go:112]: should we just use the deploy id that's generated here?

File: server/neptune/workflows/internal/deploy/revision/queue/worker.go [hunk 18]
```diff
@@ -102,14 +105,28 @@ func (w *Worker) work(ctx workflow.Context, revision string) error {
...
+	}
+	ctx = workflow.WithChildOptions(ctx, childWorkflowOptions)
+	terraformWorkflowRequest := terraform.Request{
+		Repo: w.Repo,
+		Root: w.Root,
+	}
+	err = workflow.ExecuteChildWorkflow(ctx, w.TerraformWorkflow, terraformWorkflowRequest).Get(ctx, nil)
```
Comment (nishkrishnan) [inline server/neptune/workflows/internal/deploy/revision/queue/worker.go:119]: this is a cool idea, i wonder if this will actually work cuz im not sure how the whole reflection thing even works with this.

File: server/neptune/gateway/event/push_event_handler.go [hunk 19]
```diff
@@ -80,6 +83,16 @@ func (p *PushHandler) handle(ctx context.Context, event Push) error {
...
+			Root: workflows.Root{
+				Name: "TODO",
+				Plan: workflows.Job{
+					Steps: p.generatePlanSteps(event.Repo.ID()),
+				},
+				Apply: workflows.Job{
+					Steps: p.generateApplySteps(event.Repo.ID()),
+				},
+			},
+			TerraformWorkflow: workflows.Terraform,
```
Comment (nishkrishnan) [inline server/neptune/gateway/event/push_event_handler.go:95]: requests have to be serializable since they're sent to the temporal server so i don't think this will work.  (function types are not serializable).  I'd probably just pass this in to the function directly in the workflow here if anything:
https://github.com/lyft/atlantis/blob/release-v0.17.3-lyft.1/server/neptune/workflows/deploy.go#L20

Comment (samrabelachew) [reply server/neptune/gateway/event/push_event_handler.go:95]: ah good point, forgot they need to be serialized

File: server/neptune/workflows/internal/deploy/revision/queue/worker.go (hunk 17 above)
Comment (samrabelachew) [reply server/neptune/workflows/internal/deploy/revision/queue/worker.go:112]: I wanted there to be metadata in our workflow id to make it easily recognizable when debugging, observing its process via UI, etc. What do you think?

File: server/neptune/temporalworker/server.go (hunk 16 above)
Comment (nishkrishnan) [reply server/neptune/temporalworker/server.go:129]: yeah but you can have our deploy worker as a session worker

File: server/neptune/workflows/internal/deploy/revision/queue/worker.go (hunk 17 above)
Comment (nishkrishnan) [reply server/neptune/workflows/internal/deploy/revision/queue/worker.go:112]: I'd prefer to use an obtuse identifier so its easy to like correlate things from the UI/DB/GH UI with one identifier.  debugging is made easier by using search attributes: https://docs.temporal.io/visibility/ which we should be able to leverage for metadata.

**Review by nishkrishnan — APPROVED** [review]
